connect/util.go

```diff
@@ -0,0 +1,304 @@
+		case "ST":
+			street = subjectTitle[1]
+		case "C":
+			country = subjectTitle[1]
+		case "CN":
+			appName = subjectTitle[1]
+		}
+	}
+
+	pkixName := pkix.Name{
+		Organization:       []string{org},
+		OrganizationalUnit: []string{orgUnit},
+		Locality:           []string{location},
+		StreetAddress:      []string{street},
+		Country:            []string{country},
+		CommonName:         appName,
+		Province:           []string{"Waldorf"},
+	}
+
+	// create CSR
+	var csrTemplate = x509.CertificateRequest{
+		Subject:            pkixName,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+	}
+
+	csrCertificate, err := x509.CreateCertificateRequest(rand.Reader, &csrTemplate, keys)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	csr := pem.EncodeToMemory(&pem.Block{
+		Type: "CERTIFICATE REQUEST", Bytes: csrCertificate,
+	})
+
+	var privateKey bytes.Buffer
+	err = pem.Encode(&privateKey, &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(keys)})
+
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	c.Ca.PrivateKey = privateKey.String()
+	c.Ca.Csr = string(csr)
+	return err
+}
+
+func (c *KymaConnector) GetClientCert() error {
+
+	// encode CSR to base64
+	encodedCsr := base64.StdEncoding.EncodeToString([]byte(c.Ca.Csr))
+	type Payload struct {
+		Csr string `json:"csr"`
+	}
+
+	data := Payload{
+		Csr: encodedCsr,
+	}
+	payloadBytes, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+	body := bytes.NewReader(payloadBytes)
+
+	req, err := http.NewRequest("POST", c.CsrInfo.CSRUrl, body)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+	defer resp.Body.Close()
+	certificates, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+	crtResponse := types.CRTResponse{}
+	err = json.Unmarshal(certificates, &crtResponse)
+	if err != nil {
+		return fmt.Errorf("JSON Error")
+	}
+	decodedCert, err := base64.StdEncoding.DecodeString(crtResponse.ClientCRT)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	c.Ca.PublicKey = string(decodedCert)
+	return err
+}
+
+func (c *KymaConnector) WriteClientCertificateToFile() error {
+	if c.Ca.Csr != "" {
+		err := c.StorageInterface.WriteData("generated.csr", []byte(c.Ca.Csr))
+		if err != nil {
+			return fmt.Errorf(err.Error())
+		}
+	}
+
+	if c.Ca.PublicKey != "" {
+		err := c.StorageInterface.WriteData("generated.crt", []byte(c.Ca.PublicKey))
+		if err != nil {
+			return fmt.Errorf(err.Error())
+		}
+	}
+
+	if c.Ca.PrivateKey != "" {
+		err := c.StorageInterface.WriteData("generated.key", []byte(c.Ca.PrivateKey))
+		if err != nil {
+			return fmt.Errorf(err.Error())
+		}
+	}
+	return nil
+}
+
+func (c *KymaConnector) ReadService(path string, s *Service) error {
+	path = path + ".json"
```

Comment:
ReadService is a good function name as it is not related to a specific persistence provider. Nevertheless, we should find a way to not use any file names in this context.

Reply:
Now that out interface implements separate methods for each functionality (CSR / key / config etc). we don't use the file names explicitly now, and they will be provided by the client when implementing the interface!